Example for packages with components (#3077)
conan-io/docs: examples/conanfile/package_info/components.rst

the information for each component that we want to provide to the consumers of the
game-engine package:

..  code-block:: python

    ...

    def package_info(self):
        self.cpp_info.components["algorithms"].libs = ["algorithms"]
        self.cpp_info.components["algorithms"].set_property("cmake_target_name", "algorithms")

        self.cpp_info.components["network"].libs = ["network"]
        self.cpp_info.components["network"].set_property("cmake_target_name", "network")

        self.cpp_info.components["ai"].libs = ["ai"]
        self.cpp_info.components["ai"].requires = ["algorithms"]
        self.cpp_info.components["ai"].set_property("cmake_target_name", "ai")

        self.cpp_info.components["rendering"].libs = ["rendering"]
        self.cpp_info.components["rendering"].requires = ["algorithms"]
        self.cpp_info.components["rendering"].set_property("cmake_target_name", "rendering")

There are a couple of relevant things:

- We declare the libraries generated by each of the components by setting information in
  the ``cpp_info.components`` attribute. You can set the same information for each of the

>   components as you would for the ``self.cpp_info object``. The ``cpp_info`` for

Reviewer: ```suggestion
  components as you would for the ``self.cpp_info`` object. The ``cpp_info`` for
```

  components has some defaults defined, just like it does for
  :ref:`self.cpp_info<tutorial_creating_define_package_info>`. For example, the
  ``cpp_info.components`` object provides the ``.includedirs`` and ``.libdirs`` properties

>   to define those locations, but Conan sets their value as "lib" and "include" by default,

Reviewer: ```suggestion
  to define those locations, but Conan sets their value as ``["lib"]`` and ``["include"]`` by default,
```

  target name for components like ``<package_name::component_name>``, but for this
  tutorial we will set the component target names just with the component names omitting
  the ``::``.

You can have a look at the consumer part by checking the *test_package* folder. First the
*conanfile.py*:

..  code-block:: python

    ...

    def generate(self):
        deps = CMakeDeps(self)
        deps.check_components_exist = True
        deps.generate()

You can see that we are setting the :ref:`check_components_exist
<conan_tools_cmakedeps_check_components_exist>`  property for CMakeDeps. This is not
needed, just to show how you can do if you want your consumers to fail if the component
does not exist. So, the CMakeLists.txt could look like this:

..  code-block:: text

    cmake_minimum_required(VERSION 3.15)
    project(PackageTest CXX)

    find_package(game-engine REQUIRED COMPONENTS algorithms network ai rendering)

    add_executable(example src/example.cpp)

    target_link_libraries(example algorithms 
                                  network 
                                  ai 
                                  rendering)

> And the ``find_package()`` call would fail if any of the components target does not exist. 

Reviewer: ```suggestion
And the ``find_package()`` call would fail if any of the components targets do not exist. 
```
